sub-packages/bionemo-noodles/tests/bionemo/noodles/test_nvfaidx.py

```diff
@@ -0,0 +1,403 @@
+    assert index["chr1"][0:2] == "AC"
+    assert index["chr1"][0:100000] == "ACTGACTGACTG"
+    assert index["chr2"][0:2] == "GG"
+    assert index["chr2"][0:100000] == "GGTCAAGGTCAA"
+
+    assert index["chr3"][0:2] == "AG"
+    assert index["chr3"][0:13] == "AGTCAAGGTCCAC"
+    # in progress
+    assert index["chr3"][0:14] == "AGTCAAGGTCCACG"  # adds first character from next line
+    assert index["chr3"][0:83] == "AGTCAAGGTCCACGTCAAGGTCCCGGTCAAGGTCCGTGTCAAGGTCCTAGTCAAGGTCAACGTCAAGGTCACGGTCAAGGTCA"
+    assert (
+        index["chr3"][0:84] == "AGTCAAGGTCCACGTCAAGGTCCCGGTCAAGGTCCGTGTCAAGGTCCTAGTCAAGGTCAACGTCAAGGTCACGGTCAAGGTCAG"
+    )
+    assert (
+        index["chr3"][0:10000]
+        == "AGTCAAGGTCCACGTCAAGGTCCCGGTCAAGGTCCGTGTCAAGGTCCTAGTCAAGGTCAACGTCAAGGTCACGGTCAAGGTCAG"
+    )
+    assert index["chr3"][83:84] == "G"
+
+    # Handles End of Index
+    # Full sequence
+    assert index["chr5"][0:1000000] == "A"
+    # chr5 has one char, even though this spans 2, it returns len(1)
+    assert index["chr5"][0:2] == "A"
+
+    # Handles end of multi line but non-full sequence entry
+    # Full sequence
+    assert index["chr4"][0:16] == "CCCCCCCCCCCCACGT"
+    assert index["chr4"][0:17] == "CCCCCCCCCCCCACGT"
+    assert index["chr4"][0:1000000] == "CCCCCCCCCCCCACGT"
+
+    # This one failing is bad, it means we are not calculating the newlines correctly in some conditions.
+    assert index["chr4"][0:17] == "CCCCCCCCCCCCACGT"
+
+    # Should see this is out of bounds and return empty or throw an error
+    assert index["chr4"][17:17] == ""
+
+
+def test_nvfaidx_python_interface(sample_fasta):
+    nvfaidx_index = NvFaidx(sample_fasta)
+    pyfaidx_index = pyfaidx.Fasta(sample_fasta)
+    _test_faidx_generic(nvfaidx_index)
+    _test_faidx_generic(pyfaidx_index)
+
+
+def test_pyfaidx_nvfaidx_equivalence():
+    fasta = create_test_fasta(num_seqs=2, seq_length=200000)
+    pyfaidx_fasta = pyfaidx.Fasta(fasta)
+    nvfaidx_fasta = NvFaidx(fasta)
+
+    for i in range(100):
+        # Deterministically generate regions to grab
+        seqid = f"contig{i % 2 + 1}"
+        start = i * 1000
+        end = start + 1000
+
+        if not pyfaidx_fasta[seqid][start:end] == nvfaidx_fasta[seqid][start:end]:
+            raise Exception(f"Pyfaidx and NvFaidx do not match. correct={i}")
+
+
+class TestDataset(torch.utils.data.Dataset):
+    def __init__(self, fasta_path, fasta_cls):
+        self.fasta = fasta_cls(fasta_path)
+        self.keys = list(self.fasta.keys())
+
+    def __len__(self):
+        # Gigantic, we dont care.
+        return 99999999999
+
+    def __getitem__(self, idx):
+        # Always return the same thing to keep it easy, we assume the fasta_created is doing the right thing.
+        return str(self.fasta["contig1"][150000:160000])
+
+
+@pytest.mark.skip
+@pytest.mark.xfail(reason="This is a known failure mode for pyfaidx that we are trying to prevent with nvfaidx.")
+def test_parallel_index_creation_pyfaidx():
```

Comment:
Suggest adding iteration time tests here.

Reply:
discussed offline- this is meant for the DataLoader usecase.